Changing the array behavior to start the index from 1.

hjafarpour opened the pull request:
### Description 
Currently KSQL array index starts from 0. This PR changes this behavior so the array index starts from 1 so we are compatible with SQL standard. 
This is a breaking change and we need to also include this in our release notes explicitly.
The documentation also has been updated. 

This relates to #1659 

### Testing done 
_Describe the testing stategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

hjafarpour commented:
It's ok to have breaking changes as long as we communicate them clearly. Users just need to update the queries that use arrays accordingly when they upgrade. Streams and many other projects have had breaking changes and by communicating them they don't need to keep supporting the old/deprecated behavior. If we have them as early as possible it's better.
Supporting old and new behavior in the same version would result in confusion in different ways. For instance you could force users to write their new queries based on the old/deprecated behavior even when they have the new version available. 
@big-andy-coates let me know what you think.

dguy commented:
My vote is that we make it a non-breaking change. Make a note that 0 based indexing is effectively deprecated and that people should move to 1 based indexing as in a future release 0 based indexing will be removed.

big-andy-coates commented:
> My vote is that we make it a non-breaking change. Make a note that 0 based indexing is effectively deprecated and that people should move to 1 based indexing as in a future release 0 based indexing will be removed.

You mean make the default still base-0, but have an option to switch to base-1, and encourage people to do so?

Yeah, works for me - can do the same for SUBSTRING.

You mean something like:

v5.1 - support new style, but default to old.
v5.2 - switch default to new. (could also be a major release...)
v6.0 - remove switch

big-andy-coates commented:
@hjafarpour - might also be worth documenting this in the release docs, similar to: 
https://github.com/confluentinc/docs/pull/1095

hjafarpour commented:
My main objection was having to support old behavior forever. If we only have the switch for only one more release and then remove it, it sounds good. I would suggest a process the same as @big-andy-coates described except would make the default the new behavior:

v5.1 - support new style, but default to the new (base-1).
v5.2 or 6.0 - remove the switch

What do you think @big-andy-coates and @dguy?

rodesai commented:
> My vote is that we make it a non-breaking change. Make a note that 0 based indexing is effectively deprecated and that people should move to 1 based indexing as in a future release 0 based indexing will be removed.

I don't necessarily agree with deprecating until we have an actual way of migrating queries. Currently the only option for users is to stop their queries and then start new ones. The new queries won't resume with the same state as the previously running queries. So if its a requirement for you to keep your app running then you cannot switch. So either we keep the settings/behavior around, or we deprecate and then our users can't upgrade and we have to support lots of different KSQL versions. Not sure which is worse.

Providing a migration path for this kind of change shouldn't be too hard. We just need some way to stop the query, update some settings and the query itself, and then resume it with the same consumer group. But IMO that should be a prereq for deprecating settings.

dguy commented:
Yeah, i mean leave the default as it is for now. So that queries don't break. We can then figure out if we remove it in the future.

big-andy-coates commented:
> Yeah, i mean leave the default as it is for now. So that queries don't break. We can then figure out if we remove it in the future.

Then we should also swap https://github.com/confluentinc/ksql/pull/1635 to currently default to the old school base-0 and args and document in the same way.

big-andy-coates commented:
Yep, agreed - currently fixing up the SUBSTRING change to follow the same pattern, and I've raised the following to ensure we don't forget to switch the default at in some future release: https://github.com/confluentinc/ksql/issues/1734

big-andy-coates commented:
> I don't necessarily agree with deprecating until we have an actual way of migrating queries. Currently the only option for users is to stop their queries and then start new ones. The new queries won't resume with the same state as the previously running queries. So if its a requirement for you to keep your app running then you cannot switch. So either we keep the settings/behavior around, or we deprecate and then our users can't upgrade and we have to support lots of different KSQL versions. Not sure which is worse.

@rodesai I don't see how deprecating the old way of working is going to cause issues with migrating queries - can you explain? My understanding was that for command-topic based clusters they will already have the 'legacy.mode = true' property saved in the command topic, so flipping the default to use the new more SQL-like syntax won't change any running queries.  For non-interactive mode, if they would need to add a 'set property' statement at the top of their sql file, or switch the use of SUBSTRING in the SQL file to match the new syntax, both of which would work without breaking the running app.  

However, I agree that later removing the setting all together would break old queries/apps using the old syntax. (Was this what you meant?).

> Providing a migration path for this kind of change shouldn't be too hard. We just need some way to stop the query, update some settings and the query itself, and then resume it with the same consumer group. But IMO that should be a prereq for deprecating settings.

By 'deprecating settings' to you mean 'remove settings'?

big-andy-coates commented on docs/syntax-reference.rst:
```diff
@@ -243,13 +245,15 @@ The supported column data types are:
 -  ``DOUBLE``
 -  ``VARCHAR`` (or ``STRING``)
--  ``ARRAY<ArrayType>`` (JSON and AVRO only. Index starts from 0)
+-  ``ARRAY<ArrayType>`` (JSON and AVRO only. Index starts from 1)
```
Why do we have this listed twice in the docs? Fancy pulling this out into its own section?

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java:
```diff
@@ -532,7 +532,7 @@ private String visitBooleanComparisonExpression(ComparisonExpression.Type type)
         case ARRAY:
           return new Pair<>(
-              String.format("((%s) ((%s)%s).get((int)(%s)))",
+              String.format("((%s) ((%s)%s).get(((int)(%s)) - 1))",
```
The `-1` needs to be controlled by a config setting.

Personally, I think this might be best achieved by replacing array access with a UDF.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java:
```diff
@@ -532,7 +532,7 @@ private String visitBooleanComparisonExpression(ComparisonExpression.Type type)
         case ARRAY:
           return new Pair<>(
-              String.format("((%s) ((%s)%s).get((int)(%s)))",
+              String.format("((%s) ((%s)%s).get(((int)(%s)) - 1))",
```
The `-1` needs to be controlled by a config setting.

Personally, I think this might be best achieved by replacing array access with a UDF.

hjafarpour replied:
See below comment.

big-andy-coates commented on docs/syntax-reference.rst:
```diff
@@ -243,13 +245,15 @@ The supported column data types are:
 -  ``DOUBLE``
 -  ``VARCHAR`` (or ``STRING``)
--  ``ARRAY<ArrayType>`` (JSON and AVRO only. Index starts from 0)
+-  ``ARRAY<ArrayType>`` (JSON and AVRO only. Index starts from 1)
```
Why do we have this listed twice in the docs? Fancy pulling this out into its own section?

hjafarpour replied:
Maybe in another PR since it's not directly related to this one.

big-andy-coates commented on docs/syntax-reference.rst:
```diff
@@ -166,13 +166,15 @@ The supported column data types are:
    the supported KSQL types, including the complex types ``MAP``, ``ARRAY``, and ``STRUCT``. ``STRUCT`` fields can be
    accessed in expressions using the struct dereference (``->``) operator. See :ref:`operators` for more details.
 
+Note that starting KSQL 5.1 array index starts from 1. For the versions of KSQL before 5.1, array index starts from 0.
```
I would put this 'note' inline with the ARRAY type in the list, otherwise users may miss it. (Even if that means changing the layout from a list to a table or something).

Also, need to change the default back to 0 and document the configuration setting that allows users to switch to the new more SQwhelly way of working. (as discussed above)

hjafarpour replied:
Added an inline note in addition to this paragraph. The default is now 0.

big-andy-coates commented on ksql-engine/src/test/resources/query-validation-tests/extract-json-field.json:
```diff
@@ -22,7 +22,7 @@
       "statements": [
         "CREATE STREAM TEST (array_field ARRAY<VARCHAR>) WITH (kafka_topic='test_topic', value_format='JSON');",
-        "CREATE STREAM OUTPUT AS SELECT EXTRACTJSONFIELD(array_field[0], '$.nested') AS Col1, EXTRACTJSONFIELD(array_field[1], '$.nested') AS Col2 FROM TEST;"
+        "CREATE STREAM OUTPUT AS SELECT EXTRACTJSONFIELD(array_field[1], '$.nested') AS Col1, EXTRACTJSONFIELD(array_field[2], '$.nested') AS Col2 FROM TEST;"
```
Add tests that test both old style and new style array access (with appropriate properties set), and one proving the default is the old base-zero. (See substring.json for examples).

hjafarpour replied:
Added the tests.

hjafarpour commented:
@big-andy-coates thanks for the latest review. Here is the current state:
1. I added `ksql.functions.array.legacy.base` config variable to configure the array base. If it's set to false the array index starts from 1, otherwise it starts from 0. The default is true meaning the array index starts from 0 as you suggested.
2. Updated the docs accordingly.
3. Added tests for `QueryTranslationTests` to test both cases based on the `ksql.functions.array.legacy.base` value.
4. Will update the release notes after the PR is merged.

rodesai commented on ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java:
```diff
@@ -139,7 +148,14 @@
+              KSQL_FUNCTIONS_ARRAY_LEGACY_BASE_CONFIG,
+              ConfigDef.Type.BOOLEAN,
+          true,
+          true,
```
If you set this to false we'd get the new behavior for any new queries, without breaking anything that's already running.

big-andy-coates replied:
@dguy flagging this to you - what do you reckon? I've switched SUBSTRING to default to old way, but maybe switching back to default to new way, (while maintaining backwards compatibility with old interactive queries), is the way to go?

Thoughts?

hjafarpour replied:
I like @rodesai 's suggestion too. Seems that @big-andy-coates also is leaning this way. Will wait for @dguy 's opinion before making any changes here.

dguy replied:
Sounds reasonable to me.

big-andy-coates replied:
OK, I'll switch SUBSTRING back to this model as well.

big-andy-coates replied:
SUBSTRING updated.

apurvam replied:
So then should this be updated to `false` here?

big-andy-coates replied:
> So then should this be updated to false here?

Yes, I believe so. To be clear: old should be `true` and new should be `false`.

big-andy-coates commented:
@hjafarpour what do you think to the idea of moving array access into a UDF? This would make `SqlToJavaVisitor`, which is already an overly complex class, simpler.

I think this change adds enough additional complexity to array access to warrant moving it into its own class.

Even if we don't, but think we might in the future, I would suggest renaming `ksql.functions.array.legacy.base` to `ksql.functions.arrayget.legacy.base`, as a future UDF would not be called `Array`, but something like `ArrayGet`.

hjafarpour commented:
@big-andy-coates what do you mean by making array access as a UDF? Do you mean internally or externally? Can you give an example?

big-andy-coates commented:
@hjafarpour by make it a UDF, I mean move the implementation of how array access is handled into a type, e.g. `ArrayGet`, which, given an array and an index will return the item at that index.  This `ArrayGet` can then be injected in `SqlToJavaVisitor` or we could use the re-write logic to introduce the `ArrayGet` UDF as we do for struct access.  (Basically, ArrayGet should be an internal implementation detail in the same way as `FetchFieldFromStruct` is).

The benefits of this are:
- it simplifies `SqlToJavaVisitor`
- it moves array access into a more isolated bit of code, so, for example, your new property for setting which index base to use is only in that one class.
- it allows better testing of array access
- it would mean we can more easily improve edge case handling e.g. at the moment I think if you access an array element out of bounds it will throw an exception, where as I think most other implementations just return null.

hjafarpour commented:
@big-andy-coates @apurvam there are more changes needed than you mentioned. This will include changes in multiple places including in query rewrite classes, expression type manager, function registry, and `SqlToJavaVisitor` itself. This is all for replacing the current code in `SqlToJavaVisitor` which is only `10` lines with the option of supporting both 0 based and 1 based arrays. When we deprecate 0 based array the remaining code will be 5 lines!
Plus when debugging, it's much easier to have the query as close to the original format as possible.

dguy commented:
I'm with @big-andy-coates and @apurvam on this one. The refactor seems worthwhile for all the reasons that andy already mentioned. It may be a bit of work, but delaying it just means more todo in the future as it just keeps piling up and eventually becomes so much work that it never gets done.

big-andy-coates commented:
@hjafarpour your arguments around the number of lines of code don't stand up. The same could be said for most of the existing inbuilt functions. Substring was tiny before I beefed up its error handling.  It would be possible to inline all the inbuilt functions in SqlToJavaVistor, but this would be crazy. SqlToJavaVistor is a bit of a [god class](http://wiki.c2.com/?GodClass). It should _not_ care about the inner details of how we handle array access, in the same way it shouldn't care about the inner details of how we implement Substring.  Such functionality should be factored out into separate and more targeted classes.

Also, as I've already mentioned, our current array access implementation doesn't handle out-of-bound access in the same way as other implementations. (well, the net result is the same, i.e. a null output, but for us exceptions get thrown and errors get logged). So really the implementation of array access should be larger than it currently it.

Sounds like @apurvam and @dguy are also for the refactor...

hjafarpour commented:
@big-andy-coates following my comment from our slack conversation:

> If the goal is to reduce complexity of `SqlToJavaVisitor` we can move the array and map handling code to it's own method or even a class. But I think we should try to avoid query rewriting as much as possible

I made the following change that has the benefits you mentioned without requiring UDF or query rewrite. 

- I added a new class, `ArrayGet` that provides a method for fetching array items.
- Already implements the edge case of having index out of bound and returns `null` for such cases.
- Can be tested independently.
- Isolates array access to this class.  

I believe this is far less complex than rewriting the query and using UDF. Let me know what you think!

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java:
```diff
@@ -539,11 +545,12 @@ private String visitBooleanComparisonExpression(final ComparisonExpression.Type
         case ARRAY:
           return new Pair<>(
-              String.format("((%s) ((%s)%s).get((int)(%s)))",
+              String.format("((%s) (ArrayGet.getItem(((%s) %s), (int)(%s), %s)))",
```
If index is null, then this will throw NPE.  The index may be derived from the source query e.g.

```
SELECT array_field[index_field] FROM ...
```

So we may want to consider handling a null index by just returning null. (I've not checked what other vendors do, but most seem to err on the side of not-logging-loads-of-errors and just returning null)

big-andy-coates replied:
If we throw an exception it will be logged in the processing log...

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/codegen/helpers/ArrayGet.java:
```diff
@@ -0,0 +1,34 @@
+  private ArrayGet() {}
+
+  public static Object getItem(final List<?> array, final int index, final boolean isLegacy) {
+    final int currectIndex = isLegacy ? index : index - 1;
```
nit: spelling `currect` -> `correct`?

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java:
```diff
@@ -195,6 +196,23 @@ public void shouldHandleMultiDimensionalArray() throws Exception {
+    public void shouldHandleMultiDimensionalArrayWithBaseStartingFrom1() throws Exception {
+        final KsqlConfig localKsqlConfig =
+            new KsqlConfig(Collections.singletonMap(KsqlConfig.KSQL_FUNCTIONS_ARRAY_LEGACY_BASE_CONFIG, false));
+        final CodeGenRunner localCodeGenRunner = new CodeGenRunner(schema, localKsqlConfig, functionRegistry);
```
consider making `ksqlConfig` field a mock, which means you no longer need to create a new running, just set up the mock.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java:
```diff
@@ -195,6 +196,23 @@ public void shouldHandleMultiDimensionalArray() throws Exception {
+        final List<String> innerArray1 = Arrays.asList("item_11", "item_12");
+        final List<String> innerArray2 = Arrays.asList("item_21", "item_22");
+        final Object[] args = new Object[]{Arrays.asList(innerArray1, innerArray2)};
+        final Object result = expressionEvaluatorMetadata.getExpressionEvaluator().evaluate(args);
```
consider refactoring this to be more like:

```
@Test
    public void shouldHandleMultiDimensionalArrayWithBaseStartingFrom1() throws Exception {
        // Given:
        final String simpleQuery = "SELECT col14[1][1] FROM CODEGEN_TEST;";

        final Object[] arguments = {Arrays.asList(
            Arrays.asList("item_11", "item_12"),
            Arrays.asList("item_21", "item_22"))};

        EasyMock.expect(ksqlConfig.getBoolean(KsqlConfig.KSQL_FUNCTIONS_ARRAY_LEGACY_BASE_CONFIG)).andReturn(false);
        EasyMock.replay(ksqlConfig);

        // When:
        
        final Object result = executeExpression(simpleQuery, arguments);

        // Then:
        assertThat(result, instanceOf(String.class));
        assertThat(result, equalTo("item_11"));
    }

    private Object executeExpression(final String simpleQuery, final Object[] arguments) throws Exception {
        final Analysis analysis = analyzeQuery(simpleQuery);
        final ExpressionMetadata expressionEvaluatorMetadata = codeGenRunner.buildCodeGenFromParseTree
            (analysis.getSelectExpressions().get(0));
        return expressionEvaluatorMetadata.getExpressionEvaluator().evaluate(arguments);
    }
```

To make the test more readable. The same could be applied to other tests, like the one above this one, for example.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/codegen/SqlToJavaVisitorTest.java:
```diff
@@ -75,20 +77,33 @@ public void shouldProcessBasicJavaMath() {
-    final String javaExpression = new SqlToJavaVisitor(schema, functionRegistry)
+
+  @Test
+  public void shouldProcessArrayExpressionCorrectlyForNonLegacyIndexBase() {
```
Consider refactoring to make this more readable, e.g. similar to the example I give above in the code runner class.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/codegen/helpers/ArrayGetTest.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  @Test
+  public void shouldReturnNullForOutOfBound() {
```
Add another test for bounds < 0.

hjafarpour commented on ksql-engine/src/main/java/io/confluent/ksql/codegen/helpers/ArrayGet.java:
```diff
@@ -0,0 +1,31 @@
+
+  public static Object getItem(final List<?> array, final int index, final boolean isLegacy) {
+    final int correctIndex = isLegacy ? index : index - 1;
+    return array.get(correctIndex);
```
I think for now we can throw an exception and return null. This way we make sure we don't lose information that user may need. This is the same behavior for other types of expressions. Let's make `Substring` behave the same.
I agree with the possibility of trashing the log and I think @blueedgenick 's suggestion(#1795 ) in having a data error log file would alleviate this problem.

hjafarpour commented:
@big-andy-coates we need to resolve columns and their types in analysis and planning stages and if we have a nested field we will need to implement this functionality. For fetching array items we don't need to worry about this since the array is a column and the type of the array is known.
Consider the following example:

```SELECT arrayCol[1], structCol ->field1 -> field2 FROM foo;```

When we analyze the `arrayCol`, we can directly get the type of the column but for the nested field, `structCol ->field1 -> field2`, we have to have a way to resolve the nested field and its data type. We avoid this by rewriting the nested field access as function call. This way, our existing function call analysis will take care of resolving the datatype.

miguno commented:
What's the status of this PR?

5.1 will include the change to SUBSTRING to index with start position 1 (before: 0).  So it's rather unfortunate that the same change for Arrays will not happen in 5.1, too.  Version 5.1 will then be rather confusing, because some functions start at 0, others at 1.

agavra commented:
Closing out this PR in favor of #4057
